Wrappers/Python/cil/optimisation/algorithms/PDHG.py

```diff
@@ -135,54 +129,52 @@ def update_previous_solution(self):
         tmp = self.x_old
         self.x_old = self.x
         self.x = tmp
-        tmp = self.y_old
-        self.y_old = self.y
-        self.y = tmp
-        
     def update(self):
-        
+
+        #calculate x-bar (x currently holds )
+        if self._use_axpby:
+            self.x_old.axpby((self.theta + 1.0), -self.theta , self.x, out=self.x_tmp) 
+        else:
+            self.x_old.subtract(self.x, out=self.x_tmp)
+            self.x_tmp *= self.theta
+            self.x_tmp += self.x_old
+
         # Gradient ascent for the dual variable
-        self.operator.direct(self.xbar, out=self.y_tmp)
+        self.operator.direct(self.x_tmp, out=self.y1)
         
-        # self.y_tmp *= self.sigma
-        # self.y_tmp += self.y_old
         if self._use_axpby:
-            self.y_tmp.axpby(self.sigma, 1 , self.y_old, self.y_tmp)
+            self.y1.axpby(self.sigma, 1.0 , self.y0, out=self.y1)
         else:
-            self.y_tmp *= self.sigma
-            self.y_tmp += self.y_old
+            self.y1 *= self.sigma
+            self.y1 += self.y0
+
+        self.f.proximal_conjugate(self.y1, self.sigma, out=self.y0)
 
-        # self.y = self.f.proximal_conjugate(self.y_old, self.sigma)
-        self.f.proximal_conjugate(self.y_tmp, self.sigma, out=self.y)
-        
         # Gradient descent for the primal variable
-        self.operator.adjoint(self.y, out=self.x_tmp)
-        # self.x_tmp *= -1*self.tau
-        # self.x_tmp += self.x_old
+        self.operator.adjoint(self.y0, out=self.x_tmp)
+
         if self._use_axpby:
-            self.x_tmp.axpby(-self.tau, 1. , self.x_old, self.x_tmp)
+            self.x_tmp.axpby(-self.tau, 1.0 , self.x_old, self.x_tmp)
         else:
-            self.x_tmp *= -1*self.tau
+            self.x_tmp *= -1.0*self.tau
             self.x_tmp += self.x_old
 
         self.g.proximal(self.x_tmp, self.tau, out=self.x)
-
-        # Update
-        self.x.subtract(self.x_old, out=self.xbar)
-        # self.xbar *= self.theta
-        # self.xbar += self.x
-        if self._use_axpby:
-            self.xbar.axpby(self.theta, 1 , self.x, self.xbar)
-        else:
-            self.xbar *= self.theta
-            self.xbar += self.x
-        
         
-
     def update_objective(self):
 
-        p1 = self.f(self.operator.direct(self.x)) + self.g(self.x)
-        d1 = -(self.f.convex_conjugate(self.y) + self.g.convex_conjugate(-1*self.operator.adjoint(self.y)))
+        #use y1 as temp holder
+        self.operator.direct(self.x, out=self.y1)
+        p1 = self.f(self.y1)
+        p1 += self.g(self.x)
```

Comment:
Is it possible to store the evaluation of `f` and `g` indipendently? It seems we could store at line 179 the same stuff + the bits that make the primal and dual objectives. 

There will be minor consequences for the verbose print in the the `run` method.

Reply:
I've split them in `update_objective` I don't want to add them to the array and break backward compatibility yet though. Would they be useful to view?